iroh-bytes/src/protocol.rs

```diff
@@ -517,6 +521,81 @@ impl GetRequest {
     }
 }
 
+/// A query request
+///
+/// A query request is identical to a get request, except that it describes a
+/// set of chunks for which we want availability information.
+///
+/// The answer to a query request is a set of chunks that are available, encoded
+/// as a [`RangeSpecSeq`].
+///
+/// The provider has to compute the availability information only for the requested
+/// ranges, not for the entire blob or collection. The response must contain
+/// availability information for the requested ranges, but may contain availability
+/// information for additional ranges if expedient. E.g. if the provider has
+/// a blob completely available, it may send `RangeSet2::all()` for that blob
+/// instead of limiting the response to the requested ranges.
+///
+/// However, the provider should take care to minimize the size of the response.
+/// E.g. if we have complex availability information for a range that was not
+/// requested, it should be omitted.
+///
+/// The availability information is a snapshot at one point in time.
+/// There is no guarantee that the availability won't change.
+///
+/// E.g. availability might increase when the provider is itself downloading
+/// data from other providers, or decrease when the provider is deleting data.
+#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
+pub struct QueryRequest {
+    /// blake3 hash
+    pub hash: Hash,
+    /// The range of data we request availability information for
+    ///
+    /// The first element is the parent, all subsequent elements are children.
+    pub ranges: RangeSpecSeq,
+    /// Optional Request token
+    token: Option<RequestToken>,
+}
+
+impl QueryRequest {
+    /// Request a blob or collection with specified ranges
+    pub fn new(hash: Hash, ranges: RangeSpecSeq) -> Self {
+        Self {
+            hash,
+            ranges,
+            token: None,
+        }
+    }
+
+    /// Request a collection and all its children
```

Comment:
This can also be used on a single blob?

It's a bit confusing together with `.single()` below